docs/spec_v5_errors.md

```diff
@@ -0,0 +1,173 @@
+  that callers who are iterating through a space can easily look for an
+  "out_of_range" error to detect when they are done.
+* **unimplemented**: The operation is not implemented or not
+  supported/enabled in this service.
+* **internal**: When some invariants expected by the underlying system
+  have been broken. In other words, something bad happened in the library or
+  backend service. Twirp specific issues like wire and serialization problems
+  are also reported as "internal" errors.
+* **unavailable**: The service is currently unavailable. This is most
+  likely a transient condition and may be corrected by retrying with a
+  backoff.
+* **data_loss**: Unrecoverable data loss or corruption.
+
+
+### HTTP status code
+
+Twirp services must respond with status `200` for non-error responses.
+For error responses, the staus depends on the Twirp Error Code:
+
+```
+Twirp Error Code        HTTP status code
+---------------------- -----------------------
+"canceled"            | 408 RequestTimeout
+"unknown"             | 500 Internal Server Error
+"invalid_argument"    | 400 BadRequest
+"deadline_exceeded"   | 408 RequestTimeout
+"not_found"           | 404 Not Found
+"bad_route"           | 404 Not Found
+"already_exists"      | 409 Conflict
+"permission_denied"   | 403 Forbidden
+"unauthenticated"     | 401 Unauthorized
+"resource_exhausted"  | 403 Forbidden
+"failed_precondition" | 412 Precondition Failed
+"aborted"             | 409 Conflict
+"out_of_range"        | 400 Bad Request
+"unimplemented"       | 501 Not Implemented
+"internal"            | 500 Internal Server Error
+"unavailable"         | 503 Service Unavailable
+"dataloss"            | 500 Internal Server Error
+```
+
+### Non-Twirp Errors from Intermediary Proxies
+
+It is also possible for Twirp Clients to receive non-200 responses with invalid 
+Twirp Errors (non JSON, or with invalid keys), this responses may be
+generated by proxy middleware or network issues. Twirp Clients should build a
+valid Twirp error depending on the HTTP status of those invalid responses:
+
+```
+HTTP status code           Twirp Error Code
+------------------------- -----------------------
+3xx (redirects)          | "internal"
+400 Bad Request          | "internal"
+401 Unauthorized         | "unauthenticated"
+403 Forbidden            | "permission_denied"
+404 Not Found            | "bad_route"
+429 Too Many Requests    | "unavailable"
+502 Bad Gateway          | "unavailable"
+503 Service Unavailable  | "unavailable"
+504 Gateway Timeout      | "unavailable"
+... other                | "unknown"
+```
+
+Additional metadata should be added to make it easy to identify intermediary errors:
+* `"http_error_from_intermediary": "true"`
+* `"status_code": string` (original status code on the HTTP response, e.g. `"500"`).
+* `"body": string` (original non-Twirp error response as string).
+* `"location": url-string` (only on 3xx reponses, matching the `Location` header).
```

Comment:
Definitely don't think this should be in the spec, but definitely think it should be in the docs somewhere that we do this in Go.